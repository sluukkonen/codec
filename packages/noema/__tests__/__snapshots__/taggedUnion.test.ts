import { fromJson, literal, object, taggedUnion } from "../../src/index.js"
import { expectParseFailure, expectParseSuccess } from "../helpers.js"

const Plane = object({ type: literal("plane") })
const Car = object({ type: literal("car") })
const Vehicle = taggedUnion("type", ["plane", Plane], ["car", Car])

test("should parse valid values", () => {
  expectParseSuccess(Vehicle, { type: "car" }, { type: "car" })
  expectParseSuccess(Vehicle, { type: "plane" })
})

test("should fail to parse non-objects", () => {
  expectParseFailure(Vehicle, null)
})

test("should fail to objects without the proper tag", () => {
  expectParseFailure(Vehicle, { foo: "bar" })
})

test("should fail to objects where the tag does not match", () => {
  expectParseFailure(Vehicle, { type: "boat" })
})

test("should work with complex codecs", () => {
  const complex = taggedUnion("type", [
    "complex",
    object({ type: literal("complex"), date: fromJson.date }),
  ])
  const now = new Date()

  expectParseSuccess(
    complex,
    { type: "complex", date: now.toISOString() },
    { type: "complex", date: now }
  )
})
