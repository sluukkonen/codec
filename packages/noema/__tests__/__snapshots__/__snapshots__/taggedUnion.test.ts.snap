// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail to objects where the tag does not match 1`] = `
[
  {
    "code": "invalid_tagged_union",
    "options": [
      "plane",
      "car",
    ],
    "path": [
      "type",
    ],
  },
]
`;

exports[`should fail to objects without the proper tag 1`] = `
[
  {
    "code": "invalid_tagged_union",
    "options": [
      "plane",
      "car",
    ],
    "path": [
      "type",
    ],
  },
]
`;

exports[`should fail to parse non-objects 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "path": [],
    "received": "null",
  },
]
`;
